src/Core/Services/ExecutionHelper.cs

```diff
@@ -0,0 +1,517 @@
+            {
+                string paginationObjectParentName = GetMetadataKey(context.Path) + "::" + context.Path.Parent.Depth;
+                return (IMetadata)context.ContextData[paginationObjectParentName]!;
+            }
+
+            string metadataKey = GetMetadataKey(context.Path) + "::" + context.Path.Depth;
+
+            if (context.ContextData.ContainsKey(metadataKey))
+            {
+                return (IMetadata)context.ContextData[metadataKey]!;
+            }
+            else
+            {
+                return null!;
+            }
+        }
+
+        private static IMetadata GetMetadataObjectField(IPureResolverContext context)
+        {
+            // Depth Levels:  / 0   /  1  /   2    /   3
+            // Example Path: /books/items/items[0]/publishers
+            // Depth of 1 should have key in context.ContextData
+            // Depth of 2 will not have context.ContextData entry because non-Indexer path is the path that is cached.
+            // PaginationMetadata for items will be consistent across each subitem. So we can use the same metadata for each subitem.
+            // An indexer path segment is a segment that looks like -> items[n]
+            if (context.Path.Parent is IndexerPathSegment)
+            {
+                string objectParentName = GetMetadataKey(context.Path) + "::" + context.Path.Parent!.Parent!.Depth;
+                return (IMetadata)context.ContextData[objectParentName]!;
+            }
+            else if (context.Path.Parent is not null && ((NamePathSegment)context.Path.Parent).Name != PURE_RESOLVER_CONTEXT_SUFFIX)
+            {
+                // This check handles when the current selection is a relationship field because in that case,
+                // there will be no context data entry.
+                // e.g. metadata for index 4 will not exist. only 3. 
+                // Depth: /  0   / 1  /   2    /   3      /   4
+                // Path:  /books/items/items[0]/publishers/books
+                string objectParentName = GetMetadataKey(context.Path) + "::" + context.Path.Parent!.Depth;
+                return (IMetadata)context.ContextData[objectParentName]!;
+            }
+
+            string metadataKey = GetMetadataKey(context.Path) + "::" + context.Path.Depth;
+            return (IMetadata)context.ContextData[metadataKey]!;
+        }
+
+        private static string GetMetadataKey(HotChocolate.Path path)
+        {
+            HotChocolate.Path current = path;
+
+            if (current.Parent is RootPathSegment or null)
+            {
+                return ((NamePathSegment)current).Name + PURE_RESOLVER_CONTEXT_SUFFIX;
+            }
+
+            while (current.Parent is not null)
+            {
+                current = current.Parent;
+
+                if (current.Parent is RootPathSegment or null)
+                {
+                    return ((NamePathSegment)current).Name + PURE_RESOLVER_CONTEXT_SUFFIX;
```

Comment:
seems like this can be rewritten as a recursive function

Reply:
Yeah, just rewrote and it does look more readable with less duplicate code. Downside would be callstack growth. but favoring readability here.